### PR TITLE
[ new ] Couple of useful things for Vect

### PR DESCRIPTION
**Pull request by CodingCellist**
Adds the ability to split a `Vect (k * n)` into k blocks of n or vice-versa, as well as the ability to apply a permutation to a `Vect`.

# Should this change go in the CHANGELOG?

- [x] If this is a user-facing change or a compiler change, I have updated
      `CHANGELOG.md` (and potentially also `CONTRIBUTORS.md`).

### CODE REVIEW

**Comment by CodingCellist** [comment]
These functions came up naturally as I was trying to do stuff with `Vect`s and indices. I also wanted to add:
```idris
||| All of the Fin elements as a Vect whose length is the number of elements.
|||
||| Useful, for example, when generating all the indices of a Vect.
public export
allFinsV : (n : Nat) -> Vect (S n) (Fin (S n))
allFinsV Z = FZ :: []
allFinsV (S n) = FZ :: map FS (allFinsV n)
```
But this requires importing `Data.Vect` into `Fin`, which creates a module import loop!
```
Uncaught error: Module imports form a cycle:
Decidable.Equality -> Data.DPair -> Data.Vect -> Data.Fin -> Data.List
  -> Data.Buffer -> System.File -> Control.App.FileIO -> Data.List
```
:scream:

**Comment by gallais** [comment]
I think we should have the more general:

```
allFins : (n : Nat) -> Vect n (Fin n)
```

**Comment by CodingCellist** [comment]
> I think we should have the more general:
> 
> ```
> allFins : (n : Nat) -> Vect n (Fin n)
> ```

Where should we place it though? Would putting it in `Data.Vect` avoid the cyclical import?

**Comment by buzden** [comment]
> Where should we place it though? Would putting it in `Data.Vect` avoid the cyclical import?

It definitely should

**Comment by CodingCellist** [comment]
Cool, I've added it to `Data.Vect`  : )

File: libs/base/Data/Vect.idr [hunk 1]
```diff
@@ -862,9 +899,27 @@ zipWith3IndexLinear : (0 f : _) -> (xs, ys, zs : Vect n a) -> (i : Fin n) -> ind
...
+||| @ v the vector whose elements to rearrange
+||| @ p the permutation to apply
+|||
+||| ```idris example
+||| > permute ['a', 'b', 'c', 'd'] [0, 3, 2, 1]
+||| ['a', 'd' , 'c' ,'b']
+||| ```
+export
+permute : {len : _} -> (v : Vect len a) -> (p : Vect len (Fin len)) -> Vect len a
+permute v p = zipWith index p (replicate len v)
```
Comment (gallais) [inline libs/base/Data/Vect.idr:917]: Isn't this `map (flip index v) p`? No need to keep a copy of `len` at runtime

**Comment by gallais** [comment]
I ended up pushing & merging because I needed it for #2903

**Pull request merged**